<a href="https://colab.research.google.com/github/Nikhil07-source/Machine_Learning_Algorithms/blob/main/Mercedes_Benz_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Mercedes-Benz Greener Manufacturing

<center><img src="https://static.vecteezy.com/system/resources/previews/000/063/903/non_2x/mercedes-benz-logo-vector.jpg"></center>


###Description

Reduce the time a Mercedes-Benz spends on the test bench.

###Problem Statement Scenario:
Since the first automobile, the Benz Patent Motor Car in 1886, Mercedes-Benz has stood for important automotive innovations. These include the passenger safety cell with a crumple zone, the airbag, and intelligent assistance systems. Mercedes-Benz applies for nearly 2000 patents per year, making the brand the European leader among premium carmakers. Mercedes-Benz is the leader in the premium car industry. With a huge selection of features and options, customers can choose the customized Mercedes-Benz of their dreams.

To ensure the safety and reliability of every unique car configuration before they hit the road, the company’s engineers have developed a robust testing system. As one of the world’s biggest manufacturers of premium cars, safety and efficiency are paramount on Mercedes-Benz’s production lines. However, optimizing the speed of their testing system for many possible feature combinations is complex and time-consuming without a powerful algorithmic approach.


You are required to reduce the time that cars spend on the test bench. Others will work with a dataset representing different permutations of features in a Mercedes-Benz car to predict the time it takes to pass testing. Optimal algorithms will contribute to faster testing, resulting in lower carbon dioxide emissions without reducing Mercedes-Benz’s standards.

###Following actions should be performed:

1. If for any column(s), the variance is equal to zero, then you need to remove those variable(s).
2. Check for null and unique values for test and train sets.
3. Apply label encoder.
4. Perform dimensionality reduction.
5. Predict your test_df values using XGBoost.

##Importing Pandas and Numpy Library



In [4]:
import pandas as pd #importing pandas library
import numpy as np  #importing numpy library
from sklearn.preprocessing import LabelEncoder #importing label encoder from sklearn library.
from sklearn.preprocessing import StandardScaler #importing standardscaler from sklearn library.
from sklearn.decomposition import PCA #importing Principal Component Analysis (PCA) from skelarn library.
import matplotlib.pyplot as plt #importing matplotlib library



##Importing train dataset and analysing it using pandas library.

In [5]:
df_train = pd.read_csv("https://raw.githubusercontent.com/Nikhil07-source/Datasets/refs/heads/main/Mercedes-Benz_train.csv")
df_train

ID       y  X0 X1  X2 X3 X4  X5 X6 X8  ...  X375  X376  X377  X378  \
0        0  130.81   k  v  at  a  d   u  j  o  ...     0     0     1     0   
1        6   88.53   k  t  av  e  d   y  l  o  ...     1     0     0     0   
2        7   76.26  az  w   n  c  d   x  j  x  ...     0     0     0     0   
3        9   80.62  az  t   n  f  d   x  l  e  ...     0     0     0     0   
4       13   78.02  az  v   n  f  d   h  d  n  ...     0     0     0     0   
...    ...     ...  .. ..  .. .. ..  .. .. ..  ...   ...   ...   ...   ...   
4204  8405  107.39  ak  s  as  c  d  aa  d  q  ...     1     0     0     0   
4205  8406  108.77   j  o   t  d  d  aa  h  h  ...     0     1     0     0   
4206  8412  109.22  ak  v   r  a  d  aa  g  e  ...     0     0     1     0   
4207  8415   87.48  al  r   e  f  d  aa  l  u  ...     0     0     0     0   
4208  8417  110.85   z  r  ae  c  d  aa  g  w  ...     1     0     0     0   

      X379  X380  X382  X383  X384  X385  
0        0     0     0     0     0     0  
1        0     0     0     0     0     0  
2        0     0     1     0     0     0  
3        0     0     0     0     0     0  
4        0     0     0     0     0     0  
...    ...   ...   ...   ...   ...   ...  
4204     0     0     0     0     0     0  
4205     0     0     0     0     0     0  
4206     0     0     0     0     0     0  
4207     0     0     0     0     0     0  
4208     0     0     0     0     0     0  

[4209 rows x 378 columns]

In [6]:
df_train.info()   #getting information of dataset using info function.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4209 entries, 0 to 4208
Columns: 378 entries, ID to X385
dtypes: float64(1), int64(369), object(8)
memory usage: 12.1+ MB


###Information we get using **info()** in above code.
####Following are the things we came to know by using info() function.
- In df_train dataset there are 4209 rows starting from 0 - 4209.
- There are total 378 columns starting from column ID to column X385.
- There are '1 float value', '369 integer value', '8 object' values under 378 columns.
- Memory Used in this data set is 12.1MB.

In [7]:
df_train.shape #getting the shape of the dataset which is 4209 rows with 378 columns.

(4209, 378)

In [8]:
df_train.columns #getting all the column names using ".columns".

Index(['ID', 'y', 'X0', 'X1', 'X2', 'X3', 'X4', 'X5', 'X6', 'X8',
       ...
       'X375', 'X376', 'X377', 'X378', 'X379', 'X380', 'X382', 'X383', 'X384',
       'X385'],
      dtype='object', length=378)

In [9]:
df_train.head() #getting top 5 data from our csv file which we imported.

ID       y  X0 X1  X2 X3 X4 X5 X6 X8  ...  X375  X376  X377  X378  X379  \
0   0  130.81   k  v  at  a  d  u  j  o  ...     0     0     1     0     0   
1   6   88.53   k  t  av  e  d  y  l  o  ...     1     0     0     0     0   
2   7   76.26  az  w   n  c  d  x  j  x  ...     0     0     0     0     0   
3   9   80.62  az  t   n  f  d  x  l  e  ...     0     0     0     0     0   
4  13   78.02  az  v   n  f  d  h  d  n  ...     0     0     0     0     0   

   X380  X382  X383  X384  X385  
0     0     0     0     0     0  
1     0     0     0     0     0  
2     0     1     0     0     0  
3     0     0     0     0     0  
4     0     0     0     0     0  

[5 rows x 378 columns]

In [10]:
df_train.tail() #getting last 5 data from our csv file which we imported.

ID       y  X0 X1  X2 X3 X4  X5 X6 X8  ...  X375  X376  X377  X378  \
4204  8405  107.39  ak  s  as  c  d  aa  d  q  ...     1     0     0     0   
4205  8406  108.77   j  o   t  d  d  aa  h  h  ...     0     1     0     0   
4206  8412  109.22  ak  v   r  a  d  aa  g  e  ...     0     0     1     0   
4207  8415   87.48  al  r   e  f  d  aa  l  u  ...     0     0     0     0   
4208  8417  110.85   z  r  ae  c  d  aa  g  w  ...     1     0     0     0   

      X379  X380  X382  X383  X384  X385  
4204     0     0     0     0     0     0  
4205     0     0     0     0     0     0  
4206     0     0     0     0     0     0  
4207     0     0     0     0     0     0  
4208     0     0     0     0     0     0  

[5 rows x 378 columns]

In [11]:
df_train.isnull() #checking if there is null values in dataset.

ID      y     X0     X1     X2     X3     X4     X5     X6     X8  \
0     False  False  False  False  False  False  False  False  False  False   
1     False  False  False  False  False  False  False  False  False  False   
2     False  False  False  False  False  False  False  False  False  False   
3     False  False  False  False  False  False  False  False  False  False   
4     False  False  False  False  False  False  False  False  False  False   
...     ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
4204  False  False  False  False  False  False  False  False  False  False   
4205  False  False  False  False  False  False  False  False  False  False   
4206  False  False  False  False  False  False  False  False  False  False   
4207  False  False  False  False  False  False  False  False  False  False   
4208  False  False  False  False  False  False  False  False  False  False   

      ...   X375   X376   X377   X378   X379   X380   X382   X383   X384  \
0     ...  False  False  False  False  False  False  False  False  False   
1     ...  False  False  False  False  False  False  False  False  False   
2     ...  False  False  False  False  False  False  False  False  False   
3     ...  False  False  False  False  False  False  False  False  False   
4     ...  False  False  False  False  False  False  False  False  False   
...   ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
4204  ...  False  False  False  False  False  False  False  False  False   
4205  ...  False  False  False  False  False  False  False  False  False   
4206  ...  False  False  False  False  False  False  False  False  False   
4207  ...  False  False  False  False  False  False  False  False  False   
4208  ...  False  False  False  False  False  False  False  False  False   

       X385  
0     False  
1     False  
2     False  
3     False  
4     False  
...     ...  
4204  False  
4205  False  
4206  False  
4207  False  
4208  False  

[4209 rows x 378 columns]

In [12]:
df_train.isnull().sum()       #getting total number of null values in every columns using sum() function.

ID      0
y       0
X0      0
X1      0
X2      0
       ..
X380    0
X382    0
X383    0
X384    0
X385    0
Length: 378, dtype: int64

In [13]:
df_train.describe() #getting information of our dataset using describe function.

ID            y          X10     X11          X12  \
count  4209.000000  4209.000000  4209.000000  4209.0  4209.000000   
mean   4205.960798   100.669318     0.013305     0.0     0.075077   
std    2437.608688    12.679381     0.114590     0.0     0.263547   
min       0.000000    72.110000     0.000000     0.0     0.000000   
25%    2095.000000    90.820000     0.000000     0.0     0.000000   
50%    4220.000000    99.150000     0.000000     0.0     0.000000   
75%    6314.000000   109.010000     0.000000     0.0     0.000000   
max    8417.000000   265.320000     1.000000     0.0     1.000000   

               X13          X14          X15          X16          X17  ...  \
count  4209.000000  4209.000000  4209.000000  4209.000000  4209.000000  ...   
mean      0.057971     0.428130     0.000475     0.002613     0.007603  ...   
std       0.233716     0.494867     0.021796     0.051061     0.086872  ...   
min       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
25%       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
50%       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
75%       0.000000     1.000000     0.000000     0.000000     0.000000  ...   
max       1.000000     1.000000     1.000000     1.000000     1.000000  ...   

              X375         X376         X377         X378         X379  \
count  4209.000000  4209.000000  4209.000000  4209.000000  4209.000000   
mean      0.318841     0.057258     0.314802     0.020670     0.009503   
std       0.466082     0.232363     0.464492     0.142294     0.097033   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       1.000000     0.000000     1.000000     0.000000     0.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

              X380         X382         X383         X384         X385  
count  4209.000000  4209.000000  4209.000000  4209.000000  4209.000000  
mean      0.008078     0.007603     0.001663     0.000475     0.001426  
std       0.089524     0.086872     0.040752     0.021796     0.037734  
min       0.000000     0.000000     0.000000     0.000000     0.000000  
25%       0.000000     0.000000     0.000000     0.000000     0.000000  
50%       0.000000     0.000000     0.000000     0.000000     0.000000  
75%       0.000000     0.000000     0.000000     0.000000     0.000000  
max       1.000000     1.000000     1.000000     1.000000     1.000000  

[8 rows x 370 columns]

In [14]:
df_train.select_dtypes(include=['number']).var() #getting variance of the columns whose datatype is number.

ID      5.941936e+06
y       1.607667e+02
X10     1.313092e-02
X11     0.000000e+00
X12     6.945713e-02
            ...     
X380    8.014579e-03
X382    7.546747e-03
X383    1.660732e-03
X384    4.750593e-04
X385    1.423823e-03
Length: 370, dtype: float64

##1. Removing variables whose variance is equal to zero.

In [15]:
# Identify columns with zero variance in df_train
zero_var_cols = df_train.columns[df_train.nunique() == 1]

# Drop those columns
df_train = df_train.drop(columns=zero_var_cols)

# Print the removed columns
print(f"Removed {len(zero_var_cols)} zero-variance columns:")
print(zero_var_cols.tolist())

Removed 12 zero-variance columns:
['X11', 'X93', 'X107', 'X233', 'X235', 'X268', 'X289', 'X290', 'X293', 'X297', 'X330', 'X347']


##2. Check for null and unique values for test and train sets
####We already uploaded our train data is in df_train variable, now we will upload test dataset in df_test.

In [16]:
df_train # Our train data.

ID       y  X0 X1  X2 X3 X4  X5 X6 X8  ...  X375  X376  X377  X378  \
0        0  130.81   k  v  at  a  d   u  j  o  ...     0     0     1     0   
1        6   88.53   k  t  av  e  d   y  l  o  ...     1     0     0     0   
2        7   76.26  az  w   n  c  d   x  j  x  ...     0     0     0     0   
3        9   80.62  az  t   n  f  d   x  l  e  ...     0     0     0     0   
4       13   78.02  az  v   n  f  d   h  d  n  ...     0     0     0     0   
...    ...     ...  .. ..  .. .. ..  .. .. ..  ...   ...   ...   ...   ...   
4204  8405  107.39  ak  s  as  c  d  aa  d  q  ...     1     0     0     0   
4205  8406  108.77   j  o   t  d  d  aa  h  h  ...     0     1     0     0   
4206  8412  109.22  ak  v   r  a  d  aa  g  e  ...     0     0     1     0   
4207  8415   87.48  al  r   e  f  d  aa  l  u  ...     0     0     0     0   
4208  8417  110.85   z  r  ae  c  d  aa  g  w  ...     1     0     0     0   

      X379  X380  X382  X383  X384  X385  
0        0     0     0     0     0     0  
1        0     0     0     0     0     0  
2        0     0     1     0     0     0  
3        0     0     0     0     0     0  
4        0     0     0     0     0     0  
...    ...   ...   ...   ...   ...   ...  
4204     0     0     0     0     0     0  
4205     0     0     0     0     0     0  
4206     0     0     0     0     0     0  
4207     0     0     0     0     0     0  
4208     0     0     0     0     0     0  

[4209 rows x 366 columns]

In [17]:
df_test = pd.read_csv('https://raw.githubusercontent.com/Nikhil07-source/Datasets/refs/heads/main/Mercedes-Benz_test.csv')
df_test # Our test data.

ID  X0  X1  X2 X3 X4  X5 X6 X8  X10  ...  X375  X376  X377  X378  \
0        1  az   v   n  f  d   t  a  w    0  ...     0     0     0     1   
1        2   t   b  ai  a  d   b  g  y    0  ...     0     0     1     0   
2        3  az   v  as  f  d   a  j  j    0  ...     0     0     0     1   
3        4  az   l   n  f  d   z  l  n    0  ...     0     0     0     1   
4        5   w   s  as  c  d   y  i  m    0  ...     1     0     0     0   
...    ...  ..  ..  .. .. ..  .. .. ..  ...  ...   ...   ...   ...   ...   
4204  8410  aj   h  as  f  d  aa  j  e    0  ...     0     0     0     0   
4205  8411   t  aa  ai  d  d  aa  j  y    0  ...     0     1     0     0   
4206  8413   y   v  as  f  d  aa  d  w    0  ...     0     0     0     0   
4207  8414  ak   v  as  a  d  aa  c  q    0  ...     0     0     1     0   
4208  8416   t  aa  ai  c  d  aa  g  r    0  ...     1     0     0     0   

      X379  X380  X382  X383  X384  X385  
0        0     0     0     0     0     0  
1        0     0     0     0     0     0  
2        0     0     0     0     0     0  
3        0     0     0     0     0     0  
4        0     0     0     0     0     0  
...    ...   ...   ...   ...   ...   ...  
4204     0     0     0     0     0     0  
4205     0     0     0     0     0     0  
4206     0     0     0     0     0     0  
4207     0     0     0     0     0     0  
4208     0     0     0     0     0     0  

[4209 rows x 377 columns]

In [18]:
#Checking for null values in train data.
df_train.isnull()

ID      y     X0     X1     X2     X3     X4     X5     X6     X8  \
0     False  False  False  False  False  False  False  False  False  False   
1     False  False  False  False  False  False  False  False  False  False   
2     False  False  False  False  False  False  False  False  False  False   
3     False  False  False  False  False  False  False  False  False  False   
4     False  False  False  False  False  False  False  False  False  False   
...     ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
4204  False  False  False  False  False  False  False  False  False  False   
4205  False  False  False  False  False  False  False  False  False  False   
4206  False  False  False  False  False  False  False  False  False  False   
4207  False  False  False  False  False  False  False  False  False  False   
4208  False  False  False  False  False  False  False  False  False  False   

      ...   X375   X376   X377   X378   X379   X380   X382   X383   X384  \
0     ...  False  False  False  False  False  False  False  False  False   
1     ...  False  False  False  False  False  False  False  False  False   
2     ...  False  False  False  False  False  False  False  False  False   
3     ...  False  False  False  False  False  False  False  False  False   
4     ...  False  False  False  False  False  False  False  False  False   
...   ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
4204  ...  False  False  False  False  False  False  False  False  False   
4205  ...  False  False  False  False  False  False  False  False  False   
4206  ...  False  False  False  False  False  False  False  False  False   
4207  ...  False  False  False  False  False  False  False  False  False   
4208  ...  False  False  False  False  False  False  False  False  False   

       X385  
0     False  
1     False  
2     False  
3     False  
4     False  
...     ...  
4204  False  
4205  False  
4206  False  
4207  False  
4208  False  

[4209 rows x 366 columns]

In [19]:
#Checking for null values in test data.
df_test.isnull()

ID     X0     X1     X2     X3     X4     X5     X6     X8    X10  \
0     False  False  False  False  False  False  False  False  False  False   
1     False  False  False  False  False  False  False  False  False  False   
2     False  False  False  False  False  False  False  False  False  False   
3     False  False  False  False  False  False  False  False  False  False   
4     False  False  False  False  False  False  False  False  False  False   
...     ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
4204  False  False  False  False  False  False  False  False  False  False   
4205  False  False  False  False  False  False  False  False  False  False   
4206  False  False  False  False  False  False  False  False  False  False   
4207  False  False  False  False  False  False  False  False  False  False   
4208  False  False  False  False  False  False  False  False  False  False   

      ...   X375   X376   X377   X378   X379   X380   X382   X383   X384  \
0     ...  False  False  False  False  False  False  False  False  False   
1     ...  False  False  False  False  False  False  False  False  False   
2     ...  False  False  False  False  False  False  False  False  False   
3     ...  False  False  False  False  False  False  False  False  False   
4     ...  False  False  False  False  False  False  False  False  False   
...   ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
4204  ...  False  False  False  False  False  False  False  False  False   
4205  ...  False  False  False  False  False  False  False  False  False   
4206  ...  False  False  False  False  False  False  False  False  False   
4207  ...  False  False  False  False  False  False  False  False  False   
4208  ...  False  False  False  False  False  False  False  False  False   

       X385  
0     False  
1     False  
2     False  
3     False  
4     False  
...     ...  
4204  False  
4205  False  
4206  False  
4207  False  
4208  False  

[4209 rows x 377 columns]

In [20]:
df_train[df_train.isnull().any(axis=1)] # Identify and extract rows from df_train that contain at least one null (NaN) value.

Empty DataFrame
Columns: [ID, y, X0, X1, X2, X3, X4, X5, X6, X8, X10, X12, X13, X14, X15, X16, X17, X18, X19, X20, X21, X22, X23, X24, X26, X27, X28, X29, X30, X31, X32, X33, X34, X35, X36, X37, X38, X39, X40, X41, X42, X43, X44, X45, X46, X47, X48, X49, X50, X51, X52, X53, X54, X55, X56, X57, X58, X59, X60, X61, X62, X63, X64, X65, X66, X67, X68, X69, X70, X71, X73, X74, X75, X76, X77, X78, X79, X80, X81, X82, X83, X84, X85, X86, X87, X88, X89, X90, X91, X92, X94, X95, X96, X97, X98, X99, X100, X101, X102, X103, ...]
Index: []

[0 rows x 366 columns]

In [21]:
df_test[df_test.isnull().any(axis=1)] # Identify and extract rows from df_test that contain at least one null (NaN) value.

Empty DataFrame
Columns: [ID, X0, X1, X2, X3, X4, X5, X6, X8, X10, X11, X12, X13, X14, X15, X16, X17, X18, X19, X20, X21, X22, X23, X24, X26, X27, X28, X29, X30, X31, X32, X33, X34, X35, X36, X37, X38, X39, X40, X41, X42, X43, X44, X45, X46, X47, X48, X49, X50, X51, X52, X53, X54, X55, X56, X57, X58, X59, X60, X61, X62, X63, X64, X65, X66, X67, X68, X69, X70, X71, X73, X74, X75, X76, X77, X78, X79, X80, X81, X82, X83, X84, X85, X86, X87, X88, X89, X90, X91, X92, X93, X94, X95, X96, X97, X98, X99, X100, X101, X102, ...]
Index: []

[0 rows x 377 columns]

There are no null values in Train or Test dataset, now we will check for unique values in both train & test dataset.

In [22]:
df_train.nunique() #getting unique values in train dataset.

ID      4209
y       2545
X0        47
X1        27
X2        44
        ... 
X380       2
X382       2
X383       2
X384       2
X385       2
Length: 366, dtype: int64

In [23]:
df_test.nunique() #getting unique values in test dataset.

ID      4209
X0        49
X1        27
X2        45
X3         7
        ... 
X380       2
X382       2
X383       2
X384       2
X385       2
Length: 377, dtype: int64

In [24]:
# Find columns that are in test but not in train
extra_test_columns = set(df_test.columns) - set(df_train.columns)

# Print the unexpected columns
print("Columns in test but not in train:", extra_test_columns)

Columns in test but not in train: {'X233', 'X107', 'X330', 'X293', 'X347', 'X235', 'X93', 'X290', 'X268', 'X289', 'X11', 'X297'}


In [25]:
# Drop extra columns from test data to match train data
df_test = df_test.drop(columns=['X347', 'X297', 'X293', 'X107', 'X290', 'X93', 'X330', 'X11', 'X289', 'X235', 'X233', 'X268'])

# Verify that test data now has the same columns as train data
print("Test dataset columns after dropping extras:", df_test.columns)

Test dataset columns after dropping extras: Index(['ID', 'X0', 'X1', 'X2', 'X3', 'X4', 'X5', 'X6', 'X8', 'X10',
       ...
       'X375', 'X376', 'X377', 'X378', 'X379', 'X380', 'X382', 'X383', 'X384',
       'X385'],
      dtype='object', length=365)


##3. Applying Label Encoder in Train & Test dataset.

In [26]:
# Identify categorical columns
categorical_cols = df_train.select_dtypes(include=['object']).columns
print("Categorical columns:", categorical_cols.tolist())

Categorical columns: ['X0', 'X1', 'X2', 'X3', 'X4', 'X5', 'X6', 'X8']


In [27]:
label_encoders = {}  # Dictionary to store encoders

for col in categorical_cols:
    le = LabelEncoder()
    df_train[col] = le.fit_transform(df_train[col])  # Fit & transform training data
    df_test[col] = df_test[col].map(lambda x: le.transform([x])[0] if x in le.classes_ else -1)  # Handle unseen categories
    label_encoders[col] = le  # Store encoder for future use

In [28]:
df_train.head()

ID       y  X0  X1  X2  X3  X4  X5  X6  X8  ...  X375  X376  X377  X378  \
0   0  130.81  32  23  17   0   3  24   9  14  ...     0     0     1     0   
1   6   88.53  32  21  19   4   3  28  11  14  ...     1     0     0     0   
2   7   76.26  20  24  34   2   3  27   9  23  ...     0     0     0     0   
3   9   80.62  20  21  34   5   3  27  11   4  ...     0     0     0     0   
4  13   78.02  20  23  34   5   3  12   3  13  ...     0     0     0     0   

   X379  X380  X382  X383  X384  X385  
0     0     0     0     0     0     0  
1     0     0     0     0     0     0  
2     0     0     1     0     0     0  
3     0     0     0     0     0     0  
4     0     0     0     0     0     0  

[5 rows x 366 columns]

In [29]:
df_test.head()

ID  X0  X1  X2  X3  X4  X5  X6  X8  X10  ...  X375  X376  X377  X378  X379  \
0   1  20  23  34   5   3  -1   0  22    0  ...     0     0     0     1     0   
1   2  40   3   7   0   3  -1   6  24    0  ...     0     0     1     0     0   
2   3  20  23  16   5   3  -1   9   9    0  ...     0     0     0     1     0   
3   4  20  13  34   5   3  -1  11  13    0  ...     0     0     0     1     0   
4   5  43  20  16   2   3  28   8  12    0  ...     1     0     0     0     0   

   X380  X382  X383  X384  X385  
0     0     0     0     0     0  
1     0     0     0     0     0  
2     0     0     0     0     0  
3     0     0     0     0     0  
4     0     0     0     0     0  

[5 rows x 365 columns]

## 4. Performing dimensionality reduction

In [30]:
# Exclude target column ('y') and apply standard scaling
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(df_train.drop(columns=['y']))
X_test_scaled = scaler.transform(df_test)

# Confirm the data is scaled properly
print("Shape of X_train_scaled:", X_train_scaled.shape)
print("Shape of X_test_scaled:", X_test_scaled.shape)

Shape of X_train_scaled: (4209, 365)
Shape of X_test_scaled: (4209, 365)


In [31]:
# Define the number of principal components to keep
n_components = 12
pca = PCA(n_components=n_components)

# Fit PCA on training data and transform both train & test sets
X_train_pca = pca.fit_transform(X_train_scaled)
X_test_pca = pca.transform(X_test_scaled)

# Print the new shape after PCA
print(f"Reduced train data shape: {X_train_pca.shape}")
print(f"Reduced test data shape: {X_test_pca.shape}")

Reduced train data shape: (4209, 12)
Reduced test data shape: (4209, 12)


## 5.Predict your test_df values using XGBoost.

In [32]:
from xgboost import XGBRegressor

# Initialize the model
model = XGBRegressor(n_estimators=100, learning_rate=0.1, max_depth=6, random_state=42)

# Train the model using PCA-transformed features
model.fit(X_train_pca, df_train['y'])  # Target column is 'y'

XGBRegressor(base_score=None, booster=None, callbacks=None,
             colsample_bylevel=None, colsample_bynode=None,
             colsample_bytree=None, device=None, early_stopping_rounds=None,
             enable_categorical=False, eval_metric=None, feature_types=None,
             gamma=None, grow_policy=None, importance_type=None,
             interaction_constraints=None, learning_rate=0.1, max_bin=None,
             max_cat_threshold=None, max_cat_to_onehot=None,
             max_delta_step=None, max_depth=6, max_leaves=None,
             min_child_weight=None, missing=nan, monotone_constraints=None,
             multi_strategy=None, n_estimators=100, n_jobs=None,
             num_parallel_tree=None, random_state=42, ...)

In [33]:
y_pred = model.predict(X_test_pca)

# Display first few predictions
print("Predicted values for test_df:", y_pred[:10])

Predicted values for test_df: [ 86.65735  127.65557   94.28315   77.82665  105.85277   94.236885
 104.640396  96.11248  117.06945   90.58295 ]
